cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractTwillProgramController.java

```diff
@@ -73,8 +73,15 @@ public void run() {
           // Service was killed
           stop();
         } else {
-          // Service completed by itself. Simply signal the state change of this controller.
-          complete();
+          try {
+            // This never blocks since the twill controller is already terminated. It will throw exception if
```

Comment:
Wasn't twill failure being handled before (at least most cases)?
For instance, if workflow throws exception from program code - doesn't that mean that the twill program failed?